<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>seqtrie</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">seqtrie</h1>



<div id="basic-usage" class="section level3">
<h3>Basic usage</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">dist_search</span>(x, y, <span class="at">max_distance =</span> <span class="dv">2</span>, <span class="at">nthreads =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>The above code will find all similar sequences/strings between
<code>x</code> and <code>y</code>. This will generally be significantly
faster than calculating pairwise distance or pairwise alignment.</p>
</div>
<div id="background" class="section level3">
<h3>Background</h3>
<p><code>seqtrie</code> is a collection of Radix Tree algorithms. These
include some classic algorithms like prefix lookup and more
bioinformatics focused algorithms such as alignment algorithms and
sequence distance calculations (Hamming and Levenshtein distances).</p>
<p>A trie (aka Prefix Tree) is a data structure used in many
applications. It is a space efficient data structure where a collection
of sequences are stored in a tree structure, where each leaf represents
one sequence, and each node holds one character representing a shared
prefix of all sequences descending from it. A Radix Tree (aka Compact
Prefix Tree) is an improvement on a trie, using less memory and being
generally faster. In a Radix Tree, each node is able to represent
multiple characters instead of just one.</p>
<p>Tries and Radix Trees have similar complexity to a hashmap. Storing a
sequence within the tree or looking it up are O(k) where k is the length
of the sequence. The main advantage of a trie is that it can be used to
quickly find similar sequences by various algorithms and metrics,
whereas a hashmap does not contain any sequence similarity
information.</p>
<p>See also: <a href="https://en.wikipedia.org/wiki/Radix_tree" class="uri">https://en.wikipedia.org/wiki/Radix_tree</a></p>
<p>In <code>seqtrie</code>, there are two <code>R6</code> classes:</p>
<p><code>RadixTree</code> is the primary class in this package. There
are three main methods. The <code>$insert()</code> method is used to
store sequences on the tree, <code>$erase()</code> for erasing sequences
from the tree and <code>$search()</code> for finding similar sequences
stored on the tree.</p>
<p>The second <code>R6</code> class is <code>RadixForest</code>, a
derivative data structure where separate trees are constructed for each
sequence length. This data structure has advantages and disadvantages,
discussed later.</p>
<p>Finally, there’s a simple convenience function
<code>dist_search</code> to find similar sequences using a
<code>RadixTree</code> or <code>RadixForest</code> object. This is a
wrapper around the <code>$new</code>, <code>$insert</code> and
<code>$search()</code> methods.</p>
</div>
<div id="install" class="section level3">
<h3>Install</h3>
<p><code>install.packages(&quot;seqtrie&quot;)</code></p>
</div>
<div id="more-examples" class="section level3">
<h3>More examples</h3>
<p>Below is a simple example where we insert some sequences (strings),
erase one and then plot out the tree.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>tree <span class="ot">&lt;-</span> RadixTree<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(<span class="fu">c</span>(<span class="st">&quot;cargo&quot;</span>, <span class="st">&quot;cart&quot;</span>, <span class="st">&quot;carburetor&quot;</span>, <span class="st">&quot;carbuncle&quot;</span>, <span class="st">&quot;bar&quot;</span>, <span class="st">&quot;zebra&quot;</span>))</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">erase</span>(<span class="st">&quot;zebra&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># tree$graph requires igraph package</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>); tree<span class="sc">$</span><span class="fu">graph</span>()</span></code></pre></div>
</div>
<div id="levenshtein-edit-distance-search" class="section level3">
<h3>Levenshtein “edit distance” search</h3>
<p>Below is an example using COVID19 T-cell data from Adaptive
Biotechnologies. (doi: 10.21203/rs.3.rs-51964/v1. This data is licensed
under CC 4.0.)</p>
<p>Here, we find highly similar sequences within a fixed edit distance.
For the purpose of this vignette, we sample a small selection of
sequences.</p>
<p>On the full dataset, if you tried to calculate an edit distance
matrix, it would take a pretty long time, not to mention requiring a lot
of memory. A trie-based method could be used to find similar sequences
in a fraction of the time. Approximate running times on the full dataset
using 8 threads are listed in the comments.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># 130,000 &quot;CDR3&quot; sequences</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">data</span>(covid_cdr3)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>covid_cdr3 <span class="ot">&lt;-</span> <span class="fu">sample</span>(covid_cdr3, <span class="dv">1000</span>)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>tree <span class="ot">&lt;-</span> RadixTree<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(covid_cdr3)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># Full data: 1 min</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_distance=</span><span class="dv">2</span>, <span class="at">mode=</span><span class="st">&quot;levenshtein&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co"># Alternatively, instead of using the RadixTree object directly, you can use the</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co"># dist_search function, which is a wrapper around the RadixTree object.</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">dist_search</span>(covid_cdr3, covid_cdr3, <span class="at">max_distance=</span><span class="dv">2</span>)</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co"># The output is a data.frame mapping query (search sequences)</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co"># and target (sequences inserted into the tree).</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(results, query <span class="sc">!=</span> target)</span></code></pre></div>
<pre><code>##                                           query
## 1    TGTGCCAGCAGTCACGGAACTAGCACAGATACGCAGTATTTT
## 2          TGCAGCGTTGATCTGCCGGGAGAGACCCAGTACTTC
## 3 TGTGCCAGTACTATGGGACAGGGGATGAACACTGAAGCTTTCTTT
## 4 TGTGCCAGTAGTATGGGACAGGGAATGAACACTGAAGCTTTCTTT
## 5    TGTGCCAGCAGTGACAGAACTAGCACAGATACGCAGTATTTT
## 6          TGCAGCGTTGATCTGACGGGAGAGACCCAGTACTTC
##                                          target distance
## 1    TGTGCCAGCAGTGACAGAACTAGCACAGATACGCAGTATTTT        2
## 2          TGCAGCGTTGATCTGACGGGAGAGACCCAGTACTTC        1
## 3 TGTGCCAGTAGTATGGGACAGGGAATGAACACTGAAGCTTTCTTT        2
## 4 TGTGCCAGTACTATGGGACAGGGGATGAACACTGAAGCTTTCTTT        2
## 5    TGTGCCAGCAGTCACGGAACTAGCACAGATACGCAGTATTTT        2
## 6          TGCAGCGTTGATCTGCCGGGAGAGACCCAGTACTTC        1</code></pre>
</div>
<div id="search-parameters" class="section level3">
<h3>Search parameters</h3>
<p>The <code>$search()</code> function contains two mutually exclusive
parameters: <code>max_distance</code> and <code>max_fraction</code>.The
former parameter sets an absolute threshold that limits the distance
between pairs of sequences in the output, and the latter parameter sets
a threshold relative to the query sequence length.</p>
<p>The search time is monotonically increasing with the distance
threshold, logarithmically increasing with tree size and linearly
increasing with the number of query sequences and the length of each
sequence. Overall, the algorithm is significantly faster than a
pairwise/matrix edit distance calculation for finding similar sequences.
However, <em>care still needs to be taken when setting parameters for
searching a large number of sequences (~100,000+).</em></p>
<p><strong>Some additional examples using the max_fraction
parameter.</strong></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Full data: several seconds</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_fraction=</span><span class="fl">0.035</span>, <span class="at">mode=</span><span class="st">&quot;levenshtein&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co"># Full data: 1 minute</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_fraction=</span><span class="fl">0.06</span>, <span class="at">mode=</span><span class="st">&quot;levenshtein&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># Full data: 15-20 minutes</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_fraction=</span><span class="fl">0.15</span>, <span class="at">mode=</span><span class="st">&quot;levenshtein&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="hamming-distance-search" class="section level3">
<h3>Hamming distance search</h3>
<p>Hamming distance is similar to Levenshtein distance, but does not
allow insertions or deletions. Sequences must be the same length.
Because of this restriction, Hamming distance is generally a lot
faster.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Full data: 1 second</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_fraction=</span><span class="fl">0.035</span>, <span class="at">mode=</span><span class="st">&quot;hamming&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># Full data: several seconds</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_fraction=</span><span class="fl">0.06</span>, <span class="at">mode=</span><span class="st">&quot;hamming&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co"># Full data: 1.5 minutes</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>results <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_fraction=</span><span class="fl">0.15</span>, <span class="at">mode=</span><span class="st">&quot;hamming&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="anchored-alignment-searches" class="section level3">
<h3>Anchored alignment searches</h3>
<p>An anchored alignment is a form of semi-global alignment, where the
query sequence is “anchored” (global) to the beginning of both the query
and target sequences, but is semi-global in that the end of the either
the query sequence or target sequence (but not both) can be unaligned.
This type of alignment is sometimes called an “extension” alignment in
literature.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>tree <span class="ot">&lt;-</span> RadixTree<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(<span class="st">&quot;CARTON&quot;</span>)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(<span class="st">&quot;CAR&quot;</span>)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(<span class="st">&quot;CARBON&quot;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">search</span>(<span class="st">&quot;CART&quot;</span>, <span class="at">max_distance =</span> <span class="dv">0</span>, <span class="at">mode =</span> <span class="st">&quot;anchored&quot;</span>)</span></code></pre></div>
<pre><code>##   query target distance query_size target_size
## 1  CART    CAR        0          3           3
## 2  CART CARTON        0          4           4</code></pre>
<p>Because the alignment is semi-global at the end of the alignment, the
query of “CART” finds “CAR” and “CARTON” but not “CARBON” given a max
distance of 0. Additionally, the output of an anchored search also
returns the position of the query and target at the ends. Either the
query or the target must fully align, so at least one of the end
positions will be the full length of the sequence. This type of
alignment is frequently useful in biology e.g. if you are trying to
align multiple reads that are variable in length but start at the same
genomic position or primer site.</p>
</div>
<div id="custom-distance-searches-and-affine-gap-alignment" class="section level3">
<h3>Custom distance searches and affine gap alignment</h3>
<p><code>seqtrie</code> supports custom substitution costs and affine
gap penalties. Note: we are calculating distance (higher is worse) and
not alignment score (higher is better).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>tree <span class="ot">&lt;-</span> RadixTree<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(covid_cdr3)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co"># Define a custom substitution matrix. Use generate_cost_matrix for convenience.</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>cost_mat <span class="ot">&lt;-</span> <span class="fu">generate_cost_matrix</span>(<span class="st">&quot;ACGT&quot;</span>, <span class="at">match =</span> <span class="dv">0</span>, <span class="at">mismatch =</span> <span class="dv">5</span>)</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="fu">print</span>(cost_mat)</span></code></pre></div>
<pre><code>##   A C G T
## A 0 5 5 5
## C 5 0 5 5
## G 5 5 0 5
## T 5 5 5 0</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Set gap penalties via parameters (not in the matrix):</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co"># - Linear gaps: set gap_cost only</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co"># - Affine gaps: set both gap_cost and gap_open_cost</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co"># Linear example</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>results_linear <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_distance =</span> <span class="dv">8</span>,</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>                              <span class="at">mode =</span> <span class="st">&quot;global&quot;</span>,</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>                              <span class="at">cost_matrix =</span> cost_mat,</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>                              <span class="at">gap_cost =</span> <span class="dv">2</span>,</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>                              <span class="at">nthreads =</span> <span class="dv">2</span>)</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co"># Affine example</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>results_affine <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_distance =</span> <span class="dv">8</span>,</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>                              <span class="at">mode =</span> <span class="st">&quot;global&quot;</span>,</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>                              <span class="at">cost_matrix =</span> cost_mat,</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>                              <span class="at">gap_cost =</span> <span class="dv">2</span>,</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>                              <span class="at">gap_open_cost =</span> <span class="dv">5</span>,</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>                              <span class="at">nthreads =</span> <span class="dv">2</span>)</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(results_linear, query <span class="sc">!=</span> target)</span></code></pre></div>
<pre><code>##                                            query
## 1        TGTGCCAGCAGCTTAGGACAGTCCTACGAGCAGTACTTC
## 2        TGTGCCAGCAGCTTAGGACAGTCCTACGAGCAGTACTTC
## 3     TGTGCCAGCAGTCACGGAACTAGCACAGATACGCAGTATTTT
## 4           TGCAGCGTTGATCTGCCGGGAGAGACCCAGTACTTC
## 5     TGTGCCAGCAGCTCGGCGGGGTCCTACAATGAGCAGTTCTTC
## 6        TGTGCCAGCAGTTACGGGCAGTCCTACGAGCAGTACTTC
## 7        TGTGCCAGCAGTTACGGGCAGTCCTACGAGCAGTACTTC
## 8  TGTGCCAGTACTATGGGACAGGGGATGAACACTGAAGCTTTCTTT
## 9  TGTGCCAGTAGTATGGGACAGGGAATGAACACTGAAGCTTTCTTT
## 10    TGTGCCAGCAGCTCAGGGGGCTCCTACAATGAGCAGTTCTTC
## 11          TGTGCCAGCAGTTGGGGGGGCTACGAGCAGTACTTC
## 12    TGTGCCAGCAGTTCCCCTAATAGCAATCAGCCCCAGCATTTT
## 13       TGTGCCAGCAGTTTATCGGGGTCCTACGAGCAGTACTTC
## 14       TGTGCCAGCAGTTTATCGGGGTCCTACGAGCAGTACTTC
## 15 TGTGCCAGCAGCCTTAGCGGGGTGAGCACAGATACGCAGTATTTT
## 16          TGTGCCAGCAGTCCCTCAGGGGAGACCCAGTACTTC
## 17       TGTGCCAGCAGCCTAGCAGGGGCCGGGGAGCTGTTTTTT
## 18       TGTGCCAGCAGGCCAGGACAGTCCTACGAGCAGTACTTC
## 19       TGTGCCAGCAGCTTAGAGGCGGCCGGGGAGCTGTTTTTT
## 20    TGTGCCAGCAGTCCCATAGATAGCAATCAGCCCCAGCATTTT
## 21    TGTGCCAGCAGTGACAGAACTAGCACAGATACGCAGTATTTT
## 22       TGTGCCAGCAGTTTAGGGGGTGGCTACGAGCAGTACTTC
## 23          TGTGCCAGCAGTTTCGGGGCCTACGAGCAGTACTTC
## 24    TGTGCCAGCAGCCTTAGCGGTAGCACAGATACGCAGTATTTT
## 25          TGTGCCAGCAGTCCTAGCGGCGAGACCCAGTACTTC
## 26          TGCAGCGTTGATCTGACGGGAGAGACCCAGTACTTC
##                                           target distance
## 1        TGTGCCAGCAGTTACGGGCAGTCCTACGAGCAGTACTTC        8
## 2        TGTGCCAGCAGGCCAGGACAGTCCTACGAGCAGTACTTC        8
## 3     TGTGCCAGCAGTGACAGAACTAGCACAGATACGCAGTATTTT        8
## 4           TGCAGCGTTGATCTGACGGGAGAGACCCAGTACTTC        4
## 5     TGTGCCAGCAGCTCAGGGGGCTCCTACAATGAGCAGTTCTTC        8
## 6        TGTGCCAGCAGTTTATCGGGGTCCTACGAGCAGTACTTC        8
## 7        TGTGCCAGCAGCTTAGGACAGTCCTACGAGCAGTACTTC        8
## 8  TGTGCCAGTAGTATGGGACAGGGAATGAACACTGAAGCTTTCTTT        8
## 9  TGTGCCAGTACTATGGGACAGGGGATGAACACTGAAGCTTTCTTT        8
## 10    TGTGCCAGCAGCTCGGCGGGGTCCTACAATGAGCAGTTCTTC        8
## 11       TGTGCCAGCAGTTTAGGGGGTGGCTACGAGCAGTACTTC        6
## 12    TGTGCCAGCAGTCCCATAGATAGCAATCAGCCCCAGCATTTT        8
## 13          TGTGCCAGCAGTTTCGGGGCCTACGAGCAGTACTTC        6
## 14       TGTGCCAGCAGTTACGGGCAGTCCTACGAGCAGTACTTC        8
## 15    TGTGCCAGCAGCCTTAGCGGTAGCACAGATACGCAGTATTTT        6
## 16          TGTGCCAGCAGTCCTAGCGGCGAGACCCAGTACTTC        8
## 17       TGTGCCAGCAGCTTAGAGGCGGCCGGGGAGCTGTTTTTT        8
## 18       TGTGCCAGCAGCTTAGGACAGTCCTACGAGCAGTACTTC        8
## 19       TGTGCCAGCAGCCTAGCAGGGGCCGGGGAGCTGTTTTTT        8
## 20    TGTGCCAGCAGTTCCCCTAATAGCAATCAGCCCCAGCATTTT        8
## 21    TGTGCCAGCAGTCACGGAACTAGCACAGATACGCAGTATTTT        8
## 22          TGTGCCAGCAGTTGGGGGGGCTACGAGCAGTACTTC        6
## 23       TGTGCCAGCAGTTTATCGGGGTCCTACGAGCAGTACTTC        6
## 24 TGTGCCAGCAGCCTTAGCGGGGTGAGCACAGATACGCAGTATTTT        6
## 25          TGTGCCAGCAGTCCCTCAGGGGAGACCCAGTACTTC        8
## 26          TGCAGCGTTGATCTGCCGGGAGAGACCCAGTACTTC        4</code></pre>
</div>
<div id="radix-forest-for-faster-levenshtein-searches" class="section level3">
<h3>Radix Forest for faster Levenshtein searches</h3>
<p>The <code>RadixForest</code> class is a data structure holding a
collection of Radix Trees, where a separate tree is constructed for each
sequence length. The primary advantage of <code>RadixForest</code> is
significantly faster Levenshtein searches, because you can know sequence
length up front. The disadvantages are higher memory usage, and no
support for custom distance searches.</p>
<p>Below is a brief comparison:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># RadixTree, full data: 45 seconds</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>tree <span class="ot">&lt;-</span> RadixTree<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(covid_cdr3)</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>results_tree <span class="ot">&lt;-</span> tree<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_distance=</span><span class="dv">2</span>, <span class="at">mode=</span><span class="st">&quot;levenshtein&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co"># RadixForest, full data: 19 seconds</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>frst <span class="ot">&lt;-</span> RadixForest<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>frst<span class="sc">$</span><span class="fu">insert</span>(covid_cdr3)</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>results_frst <span class="ot">&lt;-</span> frst<span class="sc">$</span><span class="fu">search</span>(covid_cdr3, <span class="at">max_distance=</span><span class="dv">2</span>, <span class="at">mode=</span><span class="st">&quot;levenshtein&quot;</span>, <span class="at">nthreads=</span><span class="dv">2</span>)</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co"># The results are the same, but order is not guaranteed</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="fu">identical</span>(</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(results_tree, query, target),</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(results_frst, query, target) )</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="finding-strings-that-start-with-a-pattern" class="section level3">
<h3>Finding strings that start with a pattern</h3>
<p>The <code>$prefix_search()</code> function can be used to find
similar sequences that start with a pattern. This is one of the classic
use cases of trie data structures, for use as a database lookup and
predictive text.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>tree <span class="ot">&lt;-</span> RadixTree<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">insert</span>(<span class="fu">c</span>(<span class="st">&quot;cargo&quot;</span>, <span class="st">&quot;cart&quot;</span>, <span class="st">&quot;carburetor&quot;</span>, <span class="st">&quot;carbuncle&quot;</span>, <span class="st">&quot;bar&quot;</span>))</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>tree<span class="sc">$</span><span class="fu">prefix_search</span>(<span class="st">&quot;car&quot;</span>)</span></code></pre></div>
<pre><code>##   query     target
## 1   car      cargo
## 2   car       cart
## 3   car carburetor
## 4   car  carbuncle</code></pre>
</div>
<div id="why-not-just-use-bowtie2-bwa-or-other-fast-alignment-software" class="section level3">
<h3>Why not just use Bowtie2, BWA or other fast alignment software?</h3>
<p>There are no apples-to-apples comparisons. With NGS alignment
software, you are looking for alignments of reads (queries)
<em>within</em> a genome reference (target). Here, we’re looking for
alignments from the query to the <em>full</em> target. However, many NGS
aligners do use Tries and similar data structures.</p>
<p>Compared to pairwise alignment packages, calculating all alignment
pairs takes much longer, but on the other hand gives you more
information.</p>
</div>
<div id="references-and-literature" class="section level3">
<h3>References and literature</h3>
<ul>
<li>“Fast string correction with Levenshtein automata” (2002) &lt;<a href="doi:10.1007/s10032-002-0082-8" class="uri">doi:10.1007/s10032-002-0082-8</a>&gt;</li>
<li>“A survey of sequence alignment algorithms for next-generation
sequencing” (2010) &lt;<a href="doi:10.1093/bib/bbq015" class="uri">doi:10.1093/bib/bbq015</a>&gt;</li>
<li>“Fast and Easy Levenshtein distance using a Trie” (2011) &lt;<a href="https://stevehanov.ca/blog/index.php?id=114" class="uri">https://stevehanov.ca/blog/index.php?id=114</a>&gt;</li>
<li>“Spell Checker Application Based on Levenshtein Automaton” (2021)
&lt;<a href="doi:10.1007/978-3-030-91608-4_5" class="uri">doi:10.1007/978-3-030-91608-4_5</a>&gt;</li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
